var/spack/repos/builtin/packages/camp/package.py

```diff
@@ -0,0 +1,46 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class Camp(CMakePackage, CudaPackage):
+    """
+    Compiler agnostic metaprogramming library providing concepts,
+    type operations and tuples for C++ and cuda
+    """
+
+    homepage = "https://github.com/LLNL/camp"
+    git      = "https://github.com/LLNL/camp.git"
+
+    version('develop', branch='master', submodules='True')
+    version('master', branch='master', submodules='True')
+    version('0.1.0', tag='v0.1.0', submodules='True')
```

Comment:
This is fine, but tags can move (even if they're not supposed to) and it is nice to keep things reproducible.  Can you either add a `commit=...` attribute, or (preferably) use a tarball and a checksum for the real releases?

I notice that there is a release here:
https://github.com/LLNL/camp/releases/tag/v0.1.0

You *can* use the `.tar.gz` link and checksum from that, but occasionally GitHub likes to change how they tar things up, and the tarballs are generated on the fly, so the `sha256` will change.  To avoid this I usually download and attach the tarball to each release of Spack (see the attached tarball here: https://github.com/spack/spack/releases/tag/v0.15.4).  That way there's a reliably checksummable URL for every release (cc: @trws).

Reply:
I pointed it to the url https://github.com/LLNL/camp/archive/v0.1.0.tar.gz instead.

I'm sorry if this seems like a simple question, but you are suggesting that I download the camp tarball manually and add it to the spack repo internally somewhere? I downloaded https://github.com/spack/spack/releases/tag/v0.15.4 and I did not see any .tar.gz files stored internally, or source code for any external projects there. I have generally been installing camp within my local version of spack. Where exactly do you want me to add the camp tarball?

Reply:
Sorry -- I wasn't very clear.  The URL you pasted above is decent, but GitHub has changed the way they tar things in the past, so I've seen the checksum of the tarballs at those links change.  This can be annoying for package managers -- see #5411 for details.

What you can do to avoid it is to download the tarball and re-upload it as an *attachment on your release*.  You go to the release, click "edit" (see below) and drop it in the upload box (see below).  That will force GitHub to keep the tarball around, and the checksum for the attachment link won't change.  As an added bonus, you'll get a download count on your release -- e.g., look at `download_count` on https://api.github.com/repos/spack/spack/releases/29655965.  It's only tracked for the attachment, not for the dynamic `.tar.gz`/`.zip` links.

<img width="1030" alt="Screen Shot 2020-10-10 at 2 53 42 PM" src="https://user-images.githubusercontent.com/299842/95665815-9c9b4f00-0b08-11eb-99f6-c2f1254c6429.png">
<img width="958" alt="Screen Shot 2020-10-10 at 2 53 57 PM" src="https://user-images.githubusercontent.com/299842/95665818-a02ed600-0b08-11eb-8e0f-b93a7364d6fc.png">

Reply:
That makes this more a note for me then, I'll make a note to do that and update the package accordingly once I add one to the camp release.